pkg/plugins/golang/v2/options.go

```diff
@@ -0,0 +1,283 @@
+	Namespaced bool
+
+	// Flags that define which parts should be scaffolded
+	DoAPI        bool
+	DoController bool
+	DoDefaulting bool
+	DoValidation bool
+	DoConversion bool
+}
+
+// Validate verifies that all the fields have valid values
+func (opts Options) Validate() error {
+	// Check that the required flags did not get a flag as their value
+	// We can safely look for a '-' as the first char as none of the fields accepts it
+	// NOTE: We must do this for all the required flags first or we may output the wrong
+	// error as flags may seem to be missing because Cobra assigned them to another flag.
+	if strings.HasPrefix(opts.Group, "-") {
+		return fmt.Errorf(groupPresent)
+	}
+	if strings.HasPrefix(opts.Version, "-") {
+		return fmt.Errorf(versionPresent)
+	}
+	if strings.HasPrefix(opts.Kind, "-") {
+		return fmt.Errorf(kindPresent)
+	}
+	// Now we can check that all the required flags are not empty
+	if len(opts.Group) == 0 {
+		return fmt.Errorf(groupRequired)
+	}
+	if len(opts.Version) == 0 {
+		return fmt.Errorf(versionRequired)
+	}
+	if len(opts.Kind) == 0 {
+		return fmt.Errorf(kindRequired)
+	}
+
+	// Check if the qualified group has a valid DNS1123 subdomain value
+	if err := validation.IsDNS1123Subdomain(opts.QualifiedGroup()); err != nil {
+		return fmt.Errorf("either group or domain is invalid: (%v)", err)
+	}
+
+	// Check if the version follows the valid pattern
+	if !versionRegex.MatchString(opts.Version) {
+		return fmt.Errorf("version must match %s (was %s)", versionPattern, opts.Version)
+	}
+
+	validationErrors := []string{}
+
+	// Require kind to start with an uppercase character
+	if string(opts.Kind[0]) == strings.ToLower(string(opts.Kind[0])) {
+		validationErrors = append(validationErrors, "kind must start with an uppercase character")
+	}
+
+	validationErrors = append(validationErrors, validation.IsDNS1035Label(strings.ToLower(opts.Kind))...)
+
+	if len(validationErrors) != 0 {
+		return fmt.Errorf("invalid Kind: %#v", validationErrors)
+	}
+
+	// TODO: validate plural strings if provided
+
+	// Ensure apiVersions for k8s types are empty or valid.
+	for typ, apiVersion := range map[string]string{
+		"CRD":     opts.CRDVersion,
+		"Webhook": opts.WebhookVersion,
+	} {
+		switch apiVersion {
+		case "", v1beta1, v1:
+		default:
+			return fmt.Errorf("%s version must be one of: v1, v1beta1", typ)
+		}
+	}
+
+	return nil
+}
+
+// QualifiedGroup returns the fully qualified group name with the available information.
+func (opts Options) QualifiedGroup() string {
+	if opts.Domain == "" {
+		return opts.Group
+	}
+	return fmt.Sprintf("%s.%s", opts.Group, opts.Domain)
+}
+
+// GVK returns the GVK identifier of a resource.
+func (opts Options) GVK(isProjectVersion2 bool) resource.GVK {
+	if isProjectVersion2 {
+		return resource.GVK{
+			Group:   opts.Group,
+			Version: opts.Version,
+			Kind:    opts.Kind,
+		}
+	}
+	return resource.GVK{
+		Group:   opts.QualifiedGroup(),
+		Version: opts.Version,
+		Kind:    opts.Kind,
```

Comment:
The same code has in more 2 diff points : https://github.com/kubernetes-sigs/kubebuilder/pull/1911/files#r545369583

Reply:
Already explained in https://github.com/kubernetes-sigs/kubebuilder/pull/1911#discussion_r545401875 why both `Options` and `Resource` need to be able to return `GVK` objects.